### DISCUSSION TITLE
How do you read a file from the filesystem?

### DISCUSSION THREAD

**Discussion by kasrasadeghi** [remix-run/remix#3846] [Q&A] [answered]
Hello!

I'm prototyping an app and I want to use a couple `.json` files as my database for now.
The normal node stuff doesn't seem to work, I want to run `await fs.readFile`.

Here are some sources I've looked through:
- https://stackoverflow.com/questions/46867517/how-to-read-file-with-async-await-properly
- https://stackoverflow.com/questions/54034648/node-js-read-file-using-async-await

**Comment by kasrasadeghi** [comment, marked as answer]
1. Create another file named `whatever.server.ts` with `export { readFile } from "node:fs/promises";`
2. In your route, do
```ts

export async function loader() {
  const fileContent = await readFile("$filepath");
  return json({content: fileContent.toString()});
}
```
where `$filepath` is relative to your project directory (or where you run your server from).

**Reply by kasrasadeghi** [reply]
Thanks to `sergiodxa` on the discord for help with this!

**Reply by Jarrodsz** [reply]
would it be possible to read from a directory and return the filenames as array?
Using https://www.geeksforgeeks.org/node-js-fs-promise-readdir-method/ perhaps?

Im not hosting on serverless or vercel so I will be able to read regular file system path.

**Reply by Jarrodsz** [reply]
I was able to get a list of files with the below code.
Im using this to load a list of mail templates.
@sergiodxa  would this be safe to use on a file based hosting environment ( digital ocean ) ?
Any suggestions on how to improve.

```
import { readdir } from "node:fs/promises";
export async function getMailTemplates() {
  const fileArray = [];
  try {
    const files = await readdir("./app/_core/mailers");
    for (const file of files) {
      fileArray.push(file);
      console.log(file);
    }
  } catch (err) {
    console.error(err);
  }
  if (fileArray.length > 0) {
    console.log(fileArray);
    return fileArray;
  }
}
```

Then in a loader im using:
```
export async function loader() {
  const mailTemplates = await getMailTemplates();
  console.log(mailTemplates);
  return json({ content: mailTemplates.toString() });
}```

**Reply by sergiocarneiro** [reply]
@Jarrodsz I believe you wanted to mention another Sérgio: @sergiodxa 🙃

**Reply by Jarrodsz** [reply]
> @Jarrodsz I believe you wanted to mention another Sérgio: @sergiodxa 🙃

My bad, my humble apologies.

**Reply by sergiodxa** [reply]
I would probably use the [`globby` package](https://www.npmjs.com/globby) to get the list of files instead of coding it myself, but the code looks mostly good, just beware that if `fileArray` is empty you're not returning anything so `getMailTemplates` returns type is an array or undefined, I think you can always return an array even if empty.
